In [1]:
from warnings import filterwarnings
filterwarnings('ignore')

### Step 1 : Read the dataset

In [2]:
import pandas as pd

df = pd.read_csv('Cars93.csv', na_values=['','NA'], keep_default_na=False)
df.head()

id Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0   1        Acura  Integra    Small       12.9   15.9       18.8        25   
1   2        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2   3         Audi       90  Compact       25.9   29.1       32.3        20   
3   4         Audi      100  Midsize       30.8   37.7       44.6        19   
4   5          BMW     535i  Midsize       23.7   30.0       36.2        22   

   MPG.highway             AirBags  ... Passengers Length  Wheelbase  Width  \
0           31                None  ...          5    177        102     68   
1           25  Driver & Passenger  ...          5    195        115     71   
2           26         Driver only  ...          5    180        102     67   
3           26                 NaN  ...          6    193        106     70   
4           30         Driver only  ...          4    186        109     69   

   Turn.circle  Rear.seat.room Luggage.room  Weight   Origin           Make  
0           37            26.5         11.0    2705  non-USA  Acura Integra  
1           38            30.0         15.0    3560  non-USA   Acura Legend  
2           37            28.0         14.0    3375  non-USA        Audi 90  
3           37            31.0         17.0    3405  non-USA       Audi 100  
4           39            27.0         13.0    3640  non-USA       BMW 535i  

[5 rows x 28 columns]

### Step 2: Perform Basic data quality checks

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 0 to 92
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  93 non-null     int64  
 1   Manufacturer        93 non-null     object 
 2   Model               93 non-null     object 
 3   Type                93 non-null     object 
 4   Min.Price           93 non-null     float64
 5   Price               93 non-null     float64
 6   Max.Price           93 non-null     float64
 7   MPG.city            93 non-null     int64  
 8   MPG.highway         93 non-null     int64  
 9   AirBags             89 non-null     object 
 10  DriveTrain          93 non-null     object 
 11  Cylinders           93 non-null     object 
 12  EngineSize          93 non-null     float64
 13  Horsepower          93 non-null     int64  
 14  RPM                 93 non-null     int64  
 15  Rev.per.mile        93 non-null     int64  
 16  Man.trans.

In [4]:
m = df.isna().sum()
m[m>0]

AirBags            4
Rear.seat.room     2
Luggage.room      11
dtype: int64

In [5]:

df.duplicated().sum()

np.int64(0)

### Step 3 : Seperate X and Y(Weight)

In [6]:
X = df.drop(columns=['id', 'Weight'])
Y = df[['Weight']]

In [7]:
X.head()

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0        Acura  Integra    Small       12.9   15.9       18.8        25   
1        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2         Audi       90  Compact       25.9   29.1       32.3        20   
3         Audi      100  Midsize       30.8   37.7       44.6        19   
4          BMW     535i  Midsize       23.7   30.0       36.2        22   

   MPG.highway             AirBags DriveTrain  ... Fuel.tank.capacity  \
0           31                None      Front  ...               13.2   
1           25  Driver & Passenger      Front  ...               18.0   
2           26         Driver only      Front  ...               16.9   
3           26                 NaN      Front  ...               21.1   
4           30         Driver only       Rear  ...               21.1   

   Passengers  Length  Wheelbase  Width Turn.circle  Rear.seat.room  \
0           5     177        102     68          37            26.5   
1           5     195        115     71          38            30.0   
2           5     180        102     67          37            28.0   
3           6     193        106     70          37            31.0   
4           4     186        109     69          39            27.0   

   Luggage.room   Origin           Make  
0          11.0  non-USA  Acura Integra  
1          15.0  non-USA   Acura Legend  
2          14.0  non-USA        Audi 90  
3          17.0  non-USA       Audi 100  
4          13.0  non-USA       BMW 535i  

[5 rows x 26 columns]

In [8]:
Y.head()

Weight
0    2705
1    3560
2    3375
3    3405
4    3640

### Step 4: Perform Preprocessing on X

In [9]:
cat = list(X.columns[X.dtypes=='object'])
con = list(X.columns[X.dtypes!='object'])

In [10]:
cat

['Manufacturer',
 'Model',
 'Type',
 'AirBags',
 'DriveTrain',
 'Cylinders',
 'Man.trans.avail',
 'Origin',
 'Make']

In [11]:
con

['Min.Price',
 'Price',
 'Max.Price',
 'MPG.city',
 'MPG.highway',
 'EngineSize',
 'Horsepower',
 'RPM',
 'Rev.per.mile',
 'Fuel.tank.capacity',
 'Passengers',
 'Length',
 'Wheelbase',
 'Width',
 'Turn.circle',
 'Rear.seat.room',
 'Luggage.room']

In [12]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

In [13]:
num_pipe = Pipeline(steps=[('impute', SimpleImputer(strategy='mean')), ('scaler', StandardScaler())])

In [14]:
cat_pipe = Pipeline(steps=[('impute',SimpleImputer(strategy='most_frequent')), ('ohe', OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [15]:
pre = ColumnTransformer([('num', num_pipe, con), ('cat', cat_pipe, cat)]).set_output(transform='pandas')

In [16]:
X_pre = pre.fit_transform(X)
X_pre

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
0        -0.485787   -0.375720       -0.282465       0.471312   
1         1.388017    1.497844        1.531409      -0.781032   
2         1.008658    0.998227        0.948052      -0.423219   
3         1.571949    1.893374        2.069191      -0.602126   
4         0.755752    1.091905        1.303535      -0.065407   
..             ...         ...             ...            ...   
88       -0.060445    0.019810        0.073018      -0.959938   
89        0.054512    0.051036        0.045673      -0.244313   
90        0.663786    0.394523        0.164167      -0.781032   
91        0.537333    0.332071        0.145937      -0.244313   
92        0.882205    0.748418        0.601684      -0.423219   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
0           0.360925        -0.841022        -0.073484  1.717489   
1          -0.770514         0.515869         1.078322  0.369586   
2          -0.581941         0.128186         0.540813  0.369586   
3          -0.581941         0.128186         0.540813  0.369586   
4           0.172352         0.806631         1.231897  0.706562   
..               ...              ...              ...       ...   
88         -1.524806        -0.162577        -0.668585 -1.315292   
89          0.172352        -0.647181        -0.188665  0.875050   
90         -0.770514         0.128186         0.655993  0.875050   
91         -0.204794        -0.356418        -0.572601  0.201098   
92         -0.204794        -0.259498         0.464025  1.549001   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
0            1.129530                -1.062184  ...                     0.0   
1            0.005661                 0.409445  ...                     0.0   
2           -0.105713                 0.072197  ...                     0.0   
3            0.410659                 1.359872  ...                     0.0   
4            0.430909                 1.359872  ...                     0.0   
..                ...                      ...  ...                     ...   
88           1.180155                 1.359872  ...                     0.0   
89           0.714407                 0.562740  ...                     0.0   
90           0.106911                 0.562740  ...                     0.0   
91          -0.237337                -0.265051  ...                     0.0   
92          -0.044963                 0.808011  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
0                       0.0                      0.0                      0.0   
1                       0.0                      0.0                      0.0   
2                       0.0                      0.0                      0.0   
3                       0.0                      0.0                      0.0   
4                       0.0                      0.0                      0.0   
..                      ...                      ...                      ...   
88                      0.0                      0.0                      0.0   
89                      0.0                      0.0                      0.0   
90                      0.0                      0.0                      0.0   
91                      0.0                      0.0                      0.0   
92                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
0                            0.0                           0.0   
1                            0.0                           0.0   
2                            0.0                           0.0   
3                            0.0                           0.0   
4                            0.0                           0.0   
..                           ...                           ...   
88                    

### Step 5: Apply train test split

In [18]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(X_pre, Y, test_size=0.2, random_state=10)

In [19]:
xtrain.head()

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
42       -0.382326   -0.209181       -0.063707       0.292406   
53       -0.715702   -0.823294       -0.874937       1.008032   
21        1.422504    1.039862        0.692834      -0.423219   
6         0.318914    0.134306       -0.018132      -0.602126   
26       -0.267368   -0.406946       -0.501224      -0.244313   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
42          0.360925        -0.453339        -0.073484  0.538074   
53          1.303790        -0.841022        -0.783765  0.369586   
21         -0.581941         0.612789         0.060893 -0.809828   
6          -0.204794         1.097393         0.502419 -0.809828   
26         -0.393368        -0.162577        -0.841356 -0.809828   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
42           0.562533                 0.102856  ...                     0.0   
53          -0.227212                -0.663618  ...                     0.0   
21          -1.108083                -0.203734  ...                     0.0   
6           -1.543455                 0.409445  ...                     0.0   
26           0.268910                -0.203734  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
42                      0.0                      0.0                      0.0   
53                      0.0                      0.0                      0.0   
21                      0.0                      0.0                      0.0   
6                       0.0                      0.0                      0.0   
26                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
42                           0.0                           0.0   
53                           0.0                           0.0   
21                           0.0                           0.0   
6                            0.0                           0.0   
26                           0.0                           0.0   

    cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  \
42                       0.0                          0.0   
53                       0.0                          0.0   
21                       0.0                          0.0   
6                        0.0                          0.0   
26                       0.0                          0.0   

    cat__Make_Volvo 240  cat__Make_Volvo 850  
42                  0.0                  0.0  
53                  0.0                  0.0  
21                  0.0                  0.0  
6                   0.0                  0.0  
26                  0.0                  0.0  

[5 rows x 257 columns]

In [20]:
xtest.head()

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
34       -0.497283   -0.573485       -0.610603       0.292406   
90        0.663786    0.394523        0.164167      -0.781032   
3         1.571949    1.893374        2.069191      -0.602126   
35       -0.301856    0.040627        0.310006      -1.317751   
19        0.146478   -0.115503       -0.318925      -0.423219   

    num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
34          0.172352        -0.647181        -0.553404  0.369586   
90         -0.770514         0.128186         0.655993  0.875050   
3          -0.581941         0.128186         0.540813  0.369586   
35         -1.713379         0.322027         0.022499 -0.809828   
19         -0.204794         0.612789         0.176074  0.032611   

    num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
34           0.015786                -0.357028  ...                     0.0   
90           0.106911                 0.562740  ...                     0.0   
3            0.410659                 1.359872  ...                     0.0   
35          -0.510711                 1.329213  ...                     0.0   
19          -0.692960                 0.409445  ...                     0.0   

    cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
34                      0.0                      0.0                      0.0   
90                      0.0                      0.0                      0.0   
3                       0.0                      0.0                      0.0   
35                      0.0                      0.0                      0.0   
19                      0.0                      0.0                      0.0   

    cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
34                           0.0                           0.0   
90                           1.0                           0.0   
3                            0.0                           0.0   
35                           0.0                           0.0   
19                           0.0                           0.0   

    cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  \
34                       0.0                          0.0   
90                       0.0                          0.0   
3                        0.0                          0.0   
35                       0.0                          0.0   
19                       0.0                          0.0   

    cat__Make_Volvo 240  cat__Make_Volvo 850  
34                  0.0                  0.0  
90                  0.0                  0.0  
3                   0.0                  0.0  
35                  0.0                  0.0  
19                  0.0                  0.0  

[5 rows x 257 columns]

In [21]:
ytrain.head()

Weight
42    3040
53    2440
21    3570
6     3470
26    3080

In [22]:
ytest.head()

Weight
34    2710
90    2810
3     3405
35    3735
19    3515

### Step 6: build and evaluate the model

#### 1. Linear Model

In [23]:
from sklearn.linear_model import LinearRegression
model1 = LinearRegression()
model1.fit(xtrain, ytrain)

LinearRegression()

In [25]:
model1.score(xtrain, ytrain)

1.0

In [26]:
model1.score(xtest, ytest)

0.8382743918509088

#### 2. Ridge Model

In [27]:
from sklearn.linear_model import Ridge
model2 = Ridge(alpha=2)
model2.fit(xtrain, ytrain)

Ridge(alpha=2)

In [28]:
model2.score(xtrain, ytrain)

0.9953183637556047

In [29]:
model2.score(xtest, ytest)

0.8656890544193268

### Tuning the alpha parameter with Grid serch cross validation 

In [30]:
from sklearn.model_selection import cross_val_score
r2_scores_ridge = cross_val_score(model2, xtrain, ytrain, cv=5, scoring='r2')
r2_scores_ridge

array([0.97433064, 0.96683336, 0.8870119 , 0.95211521, 0.93550405])

In [31]:
r2_scores_ridge.mean()

np.float64(0.9431590326564206)

In [32]:
r2_scores_linear = cross_val_score(model1, xtrain, ytrain, cv=5, scoring='r2')
r2_scores_linear

array([0.97469114, 0.96348516, 0.87554023, 0.95454911, 0.93554437])

In [33]:
r2_scores_linear.mean()

np.float64(0.9407620016166703)

### Generating alpha values 

In [34]:
import numpy as np

alphas= np.arange(start=0.1, stop=100, step=0.1)
alphas

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15

In [39]:
params = {'alpha' : alphas}
params

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [40]:
from sklearn.model_selection import GridSearchCV
model_ridge = Ridge()
gscv_ridge = GridSearchCV(model_ridge, param_grid=params, cv=5, scoring='r2')
gscv_ridge.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7....
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [41]:
gscv_ridge.best_params_

{'alpha': np.float64(3.8000000000000003)}

In [42]:
gscv_ridge.best_score_

np.float64(0.9434391769932094)

In [43]:
best_ridge = gscv_ridge.best_estimator_
best_ridge

Ridge(alpha=np.float64(3.8000000000000003))

In [44]:
best_ridge.score(xtrain, ytrain)

0.9906446755991962

In [45]:
best_ridge.score(xtest, ytest)

0.8796471048193708

#### 3. Lasso

In [46]:
from sklearn.linear_model import Lasso
model3 = Lasso(alpha=1)
model3.fit(xtrain, ytrain)

Lasso(alpha=1)

In [47]:
model3.score(xtrain, ytrain)

0.9936354478021407

In [48]:
model3.score(xtest, ytest)

0.7816304364881249

In [49]:
params

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [50]:
model_lasso = Lasso()
gscv_lasso = GridSearchCV(model_lasso, param_grid=params, cv=5, scoring='r2')
gscv_lasso.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7....
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [51]:
gscv_lasso.best_params_

{'alpha': np.float64(5.2)}

In [52]:
gscv_lasso.best_score_

np.float64(0.9422784524364886)

In [53]:
best_lasso = gscv_lasso.best_estimator_
best_lasso

Lasso(alpha=np.float64(5.2))

In [54]:
best_lasso.score(xtrain, ytrain)

0.9714666534740786

In [55]:
best_lasso.score(xtest, ytest)

0.869457987545957

In [59]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error, r2_score

In [60]:
def evaluate_model (model, x, y):
    ypred = model.predict(x)

    mse = mean_squared_error(y, ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(y, ypred)
    mape = mean_absolute_percentage_error(y, ypred)
    r2 = r2_score(y, ypred)

    print(f'Mean Squared Error :{mse:.2f}')
    print(f'Root Mean Squared Error :{rmse:.2f}')
    print(f'Mean Absolute Error :{mae:.2f}')
    print(f'Mean Absolute Percentage Error :{mape:.4f}')
    print(f'R2 score:{r2:.4f}')

In [61]:
evaluate_model(model1, xtrain, ytrain)

Mean Squared Error :0.00
Root Mean Squared Error :0.00
Mean Absolute Error :0.00
Mean Absolute Percentage Error :0.0000
R2 score:1.0000


In [62]:
evaluate_model(model1, xtest, ytest)

Mean Squared Error :37574.50
Root Mean Squared Error :193.84
Mean Absolute Error :147.27
Mean Absolute Percentage Error :0.0490
R2 score:0.8383


In [63]:
evaluate_model(best_ridge, xtrain, ytrain)

Mean Squared Error :3479.39
Root Mean Squared Error :58.99
Mean Absolute Error :44.34
Mean Absolute Percentage Error :0.0145
R2 score:0.9906


In [64]:
evaluate_model(best_ridge, xtest, ytest)

Mean Squared Error :27962.18
Root Mean Squared Error :167.22
Mean Absolute Error :130.26
Mean Absolute Percentage Error :0.0438
R2 score:0.8796


In [65]:
evaluate_model(best_lasso, xtrain, ytrain)

Mean Squared Error :10611.98
Root Mean Squared Error :103.01
Mean Absolute Error :81.72
Mean Absolute Percentage Error :0.0265
R2 score:0.9715


In [66]:
evaluate_model(best_lasso, xtest, ytest)

Mean Squared Error :30329.47
Root Mean Squared Error :174.15
Mean Absolute Error :131.92
Mean Absolute Percentage Error :0.0446
R2 score:0.8695


### After compairing above r2 scores Ridge model has highest score of 0.8796

#### Choose ridge model for out of sample predictions

In [67]:
xnew = pd.read_csv('sample.csv', na_values=['', 'NA'], keep_default_na=False)
xnew

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0         Audi      100  Midsize       30.8   37.7       44.6        19   
1      Pontiac  Sunbird  Compact        9.4   11.1       12.8        23   
2    Chevrolet   Lumina  Midsize       13.4   15.9       18.4        21   
3        Mazda     RX-7   Sporty       32.5   32.5       32.5        17   
4   Volkswagen      Fox    Small        8.7    9.1        9.5        25   

   MPG.highway      AirBags DriveTrain  ... Fuel.tank.capacity  Passengers  \
0           26          NaN      Front  ...               15.0           6   
1           31         None      Front  ...               15.2           5   
2           29         None      Front  ...               16.5           6   
3           25  Driver only       Rear  ...               20.0           2   
4           33         None      Front  ...               12.4           4   

   Length  Wheelbase  Width Turn.circle  Rear.seat.room  Luggage.room  \
0     190        106     65          37            31.0          17.0   
1     181        101     66          39            25.0          13.0   
2     198        108     71          40            28.5          16.0   
3     169         96     69          37             NaN           NaN   
4     163         93     63          34            26.0          10.0   

    Origin              Make  
0  non-USA          Audi 100  
1      USA   Pontiac Sunbird  
2      USA  Chevrolet Lumina  
3  non-USA        Mazda RX-7  
4  non-USA    Volkswagen Fox  

[5 rows x 26 columns]

In [68]:
pre

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('impute', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['Min.Price', 'Price', 'Max.Price', 'MPG.city',
                                  'MPG.highway', 'EngineSize', 'Horsepower',
                                  'RPM', 'Rev.per.mile', 'Fuel.tank.capacity',
                                  'Passengers', 'Length', 'Wheelbase', 'Width',
                                  'Turn.circle', 'Rear.seat.room',
                                  'Luggage.room']),
                                ('cat',
                                 Pipeline(steps=[('impute',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['Manufacturer', 'Model', 'Type', 'AirBags',
                                  'DriveTrain', 'Cylinders', 'Man.trans.avail',
                                  'Origin', 'Make'])])

In [69]:
xnew_pre = pre.transform(xnew)
xnew_pre

num__Min.Price  num__Price  num__Max.Price  num__MPG.city  \
0        1.571949    1.893374        2.069191      -0.602126   
1       -0.888138   -0.875337       -0.829362       0.113500   
2       -0.428309   -0.375720       -0.318925      -0.244313   
3        1.767376    1.352122        0.966282      -0.959938   
4       -0.968608   -1.083511       -1.130155       0.471312   

   num__MPG.highway  num__EngineSize  num__Horsepower  num__RPM  \
0         -0.581941         0.128186         0.540813  0.369586   
1          0.360925        -0.647181        -0.649388 -0.135877   
2         -0.016221        -0.453339        -0.649388 -0.135877   
3         -0.770514        -1.325626         2.134145  2.054464   
4          0.738071        -0.841022        -1.206095  0.369586   

   num__Rev.per.mile  num__Fuel.tank.capacity  ...  cat__Make_Toyota Camry  \
0           0.410659                -0.510323  ...                     0.0   
1           0.673908                -0.449005  ...                     0.0   
2           0.532158                -0.050439  ...                     0.0   
3          -0.014589                 1.022624  ...                     0.0   
4           0.441034                -1.307455  ...                     0.0   

   cat__Make_Toyota Celica  cat__Make_Toyota Previa  cat__Make_Toyota Tercel  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   cat__Make_Volkswagen Corrado  cat__Make_Volkswagen Eurovan  \
0                           0.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           0.0                           0.0   

   cat__Make_Volkswagen Fox  cat__Make_Volkswagen Passat  cat__Make_Volvo 240  \
0                       0.0                          0.0                  0.0   
1                       0.0                          0.0                  0.0   
2                       0.0                          0.0                  0.0   
3                       0.0                          0.0                  0.0   
4                       1.0                          0.0                  0.0   

   cat__Make_Volvo 850  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  

[5 rows x 257 columns]

In [70]:
preds = best_ridge.predict(xnew_pre)
preds

array([[3354.76382703],
       [2652.51328626],
       [3057.04172854],
       [3269.37188001],
       [2242.37830206]])

In [71]:
df_final = xnew[['Manufacturer', 'Model']]
df_final

Manufacturer    Model
0         Audi      100
1      Pontiac  Sunbird
2    Chevrolet   Lumina
3        Mazda     RX-7
4   Volkswagen      Fox

In [72]:
df_final[['Weight_pred']] = preds

df_final

Manufacturer    Model  Weight_pred
0         Audi      100  3354.763827
1      Pontiac  Sunbird  2652.513286
2    Chevrolet   Lumina  3057.041729
3        Mazda     RX-7  3269.371880
4   Volkswagen      Fox  2242.378302

In [73]:
df_final.to_csv('RidgeResults.csv', index=False)